arch/risc-v/src/common/riscv_swint.c

```diff
@@ -48,16 +48,61 @@
  * Pre-processor Definitions
  ****************************************************************************/
 
-#ifdef CONFIG_LIB_SYSCALL
-#  define TCB_FLAGS_OFFSET offsetof(struct tcb_s, flags)
-#endif
-
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
 
 #ifdef CONFIG_LIB_SYSCALL
 
+/****************************************************************************
+ * Name: do_syscall
+ *
+ * Description:
+ *   Call the stub function corresponding to the system call.  NOTE the non-
+ *   standard parameter passing:
+ *
+ *     A0 = SYS_ call number
+ *     A1 = parm0
+ *     A2 = parm1
+ *     A3 = parm2
+ *     A4 = parm3
+ *     A5 = parm4
+ *     A6 = parm5
+ *
+ ****************************************************************************/
+
+uintptr_t do_syscall(unsigned int nbr, uintptr_t parm1,
```

Comment:
Yeah can be changed to static

Reply:
Changing the function to static makes the compiler inline the function into dispatch_syscall (it was kind of expected), which causes a register context corruption. I need to look into the inline assembly code more carefully. Maybe the clobber list is incomplete.

Reply:
As I suspected, the compiler uses a7 as a local variable but does not understand it can get clobbered by the system call
(it doesn't understand the inlined assembly code does a jump..). What a pain.

I'll add noinline_function and see if that fixes the issue.